xray/segment.go

```diff
@@ -61,6 +62,13 @@ func BeginSegment(ctx context.Context, name string) (context.Context, *Segment)
 }
 
 func BeginSegmentWithSampling(ctx context.Context, name string, r *http.Request, traceHeader *header.Header) (context.Context, *Segment) {
+	// If SDK is disabled then return with an empty segment
+	disableKey := os.Getenv("AWS_XRAY_SDK_DISABLED")
+	if strings.ToLower(disableKey) == "true" {
+		seg := &Segment{}
+		return context.WithValue(ctx, ContextKey, seg), seg
+	}
+
 	seg := basicSegment(name, nil)
```

Comment:
You should either construct a basic segment or at least populate the `parentSegment` of the new segment before returning it IMO. There are several methods that assume the fields populated in `basicSegment` are non-nil, e.g. `DownstreamHeader()` in `segment_model.go`. If you return an empty segment, the `parentSegment` field will be `nil` and the method will throw a nil dereference error.

Reply:
That's a good point. After scanning the code base only DownstreamHeader() method can cause the nil pointer exception because customer might call that method directly otherwise other APIs are used by SDK internally. Also, if we don't populate segment fields then we can avoid unnecessary operations in the case of beginsubsegment. In this draft I have added disableKey logic in methods which customer can directly call to avoid throwing panic.

Reply:
Fair enough, this looks good to me!